src/prefect/_internal/compatibility/deprecated.py

```diff
@@ -0,0 +1,204 @@
+e.g. Jan 2023.
+"""
+import functools
+import warnings
+from typing import Any, Callable, Optional, Type, TypeVar
+
+import pendulum
+import pydantic
+
+from prefect.utilities.callables import get_call_parameters
+from prefect.utilities.importtools import to_qualified_name
+
+T = TypeVar("T", bound=Callable)
+M = TypeVar("M", bound=pydantic.BaseModel)
+
+
+DEPRECATED_WARNING = "{name} has been deprecated{when}. It will not be available after {end_date}. {help}"
+DEPRECATED_MOVED_WARNING = (
+    "{name} has moved to {new_location}. It will not be available at the old import "
+    "path after {end_date}. {help}"
+)
+DEPRECATED_DATEFMT = "MMM YYYY"
+
+
+class PrefectDeprecationWarning(DeprecationWarning):
+    """
+    A deprecation warning.
+    """
+
+
+def generate_deprecation_message(
+    name: str,
+    start_date: Optional[str] = None,
+    end_date: Optional[str] = None,
+    help: str = "",
+    when: str = "",
+):
+    if not start_date and not end_date:
+        raise ValueError(
+            "A start date is required if an end date is not provided. "
+            f"Suggested start date is {pendulum.now().format(DEPRECATED_DATEFMT)!r}"
+        )
+
+    if not end_date:
+        parsed_start_date = pendulum.from_format(start_date, DEPRECATED_DATEFMT)
+        parsed_end_date = parsed_start_date.add(months=6)
+        end_date = parsed_end_date.format(DEPRECATED_DATEFMT)
+    else:
+        # Validate format
+        pendulum.from_format(end_date, DEPRECATED_DATEFMT)
+
+    if when:
+        when = " when " + when
+
+    message = DEPRECATED_WARNING.format(
+        name=name, when=when, end_date=end_date, help=help
+    )
+    return message.rstrip()
+
+
+def deprecated_callable(
+    *,
+    start_date: Optional[str] = None,
+    end_date: Optional[str] = None,
+    stacklevel: int = 2,
+    help: str = "",
+) -> Callable[[T], T]:
+    def decorator(fn: T):
+        message = generate_deprecation_message(
+            name=to_qualified_name(fn),
+            start_date=start_date,
+            end_date=end_date,
+            help=help,
+        )
+
+        @functools.wraps(fn)
+        def wrapper(*args, **kwargs):
+            warnings.warn(message, PrefectDeprecationWarning, stacklevel=stacklevel)
+            return fn(*args, **kwargs)
+
+        return wrapper
+
+    return decorator
+
+
+def deprecated_parameter(
+    name: str,
+    *,
+    start_date: Optional[str] = None,
+    end_date: Optional[str] = None,
+    stacklevel: int = 2,
+    help: str = "",
+    when: Optional[Callable[[Any], bool]] = None,
+    when_message: str = "",
+) -> Callable[[T], T]:
+    """
+    Mark a parameter in a callable as deprecated.
+
+    Example:
+
+        ```python
+
+        @deprecated_parameter("y", when=lambda y: y is not None)
+        def foo(x, y = None):
+            return x + 1 + (y or 0)
+        ```
+    """
+
+    when = when or (lambda _: True)
```

Comment:
If `when` is False, would this trigger `or`?

Reply:
The deprecation warning will trigger if `when` evaluates to `True`. In this example, passing any value for `y` other than `None` will result in the deprecation warning.

Reply:
Thanks, I think I got tripped up on: `Optional[Callable[[Any], bool]]` I misread this as `Optional[Union[Callable, bool]]`

Reply:
Indeed `when` should be a callable and this sets the default callable to be a noop that always evaluates to true to simplify the downstream logic